### PR TITLE
Task/203/simple test to check bootstrap oneliner

### PR DESCRIPTION
**Pull request by svatwork**
# Pull Request

This pull request is linked to #203. The pull request has been outlined in that issue. 

## Description

I've added a Dockerfile layer with the bats image and right now it will execute the first test that is written, which is passing. 

However, we've discussed that @Morriz doesn't like to have too many dependencies in the Dockerfile so with the pull request draft I would like to receive a suggestion how to structure it to accept the pull request.

The docker layer alias is also open to discussion, so let me know what name you would like to see. 

## Motivation and Context

#203 

## Tests

`docker build --target unit-tests -t test:latest .  `

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have added tests (if applicable)
- [ ] I have updated documentation (if applicable)
- [x] My code follows the code style of this project (automated for vscode!)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -18,6 +18,23 @@ RUN [ "$SKIP_TESTS" = 'false' ] && \
   bin/validate-values.sh && \
   bin/validate-templates.sh || true
 
+#-----------------------------
+FROM bats/bats:latest as test
```
Comment (Morriz) [inline Dockerfile:22]: I prefer to add bats to our tools image and run on that, so we might catch missing deps while running tests.

**Review by Morriz — CHANGES_REQUESTED** [review]
Looking good. One remark. Can you also pull master? It has changes to the Dockerfile.

**Comment by svatwork** [comment]
Not sure what version of otomi/tools I should target, since I modified tools/Dockerfile.

**Comment by svatwork** [comment]
For this build to pass, otomi/tools should be bumped.

File: Dockerfile [hunk 2]
```diff
@@ -14,7 +14,18 @@ RUN if [ "$SKIP_TESTS" = 'false' ]; then \
...
+ENV APP_HOME=/home/app/stack
+RUN mkdir -p $APP_HOME
+WORKDIR $APP_HOME
+
+COPY . .
+
+RUN bats bin/tests
+
+#-----------------------------
+FROM otomi/tools:1.4.8 as prod
```
Comment (Morriz) [inline Dockerfile:28]: can this be 1.4.9 also then?

File: bin/tests/bootstrap-unit.bats [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+teardown () {
+    temp_del "$test_temp_dir"
+    unset ENV_DIR
+    unset env_path
+}
+
+@test "the env folder should not exist without bootstrap.sh" { 
+    assert_file_not_exist "$env_path"
+}
```
Comment (Morriz) [inline bin/tests/bootstrap-unit.bats:23]: this is not testing anything, please remove

Comment (Morriz) [reply bin/tests/bootstrap-unit.bats:23]: instead you may add a test that checks if env files don't get overwritten when they exist

File: bin/tests/bootstrap-integration.bats [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+@test "executing bootstrap.sh should fail without env folder" {
+    run bin/bootstrap.sh
+    assert_failure
+}
+
+@test "executing bootstrap.sh multiple times should pass" {
+    git init "$test_temp_dir"
+    bin/bootstrap.sh 
+    run bin/bootstrap.sh
```
Comment (Morriz) [inline bin/tests/bootstrap-integration.bats:33]: I'm curious. Why is the second one invoked with `run`?

File: bin/tests/bootstrap-integration.bats [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    assert_success
+}
```
Comment (Morriz) [inline bin/tests/bootstrap-integration.bats:35]: Why is this file considered to be running integration tests? I don't see any. Can we just keep the unit tests and add more tests there?

File: tools/Dockerfile [hunk 6]
```diff
@@ -79,6 +79,15 @@ RUN curl -sL https://deb.nodesource.com/setup_lts.x | bash -
...
   npm install -g ajv-cli
 
+# bats https://github.com/bats-core/bats-core
+RUN git clone https://github.com/bats-core/bats-core.git && \ 
+  cd bats-core && \
+  ./install.sh /usr/local && \
+  cd /usr/local/lib && \
+  git clone https://github.com/ztombol/bats-support.git && \
+  git clone https://github.com/ztombol/bats-assert.git && \ 
+  git clone https://github.com/ztombol/bats-file.git
```
Comment (Morriz) [inline tools/Dockerfile:89]: can you instead download release binaries like we are doing everywhere else?

**Review by Morriz — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (svatwork) [reply Dockerfile:28]: Yes actually I want to merge the "aws" tools image and then bump this one.

File: bin/tests/bootstrap-unit.bats (hunk 3 above)
Comment (svatwork) [reply bin/tests/bootstrap-unit.bats:23]: Well I was watching some lectures on software testing and they mentioned a testrunner should at least have a case that 1) passes, 2) fails, and 3) passes after failing. So I think this would increase test coverage in the edge case covered by the test. I just wanted to deliver this feature so we can expand upon it in the future, so I didn't spend too much time overengineering.

File: bin/tests/bootstrap-integration.bats (hunk 4 above)
Comment (svatwork) [reply bin/tests/bootstrap-integration.bats:33]: From the docs:

```
When not to use run

In some cases, using run is redundant and results in a longer and less readable code. Here are a few examples.

In case you only need to check the command succeeded, it is better to not use run, since
run command args ...
echo "$output"
[ "$status" -eq 0 ]
is equivalent to

command args ...
```

So the first run is actually not testing for completion but the second time it's run the exit code should be captured.

File: bin/tests/bootstrap-integration.bats (hunk 5 above)
Comment (svatwork) [reply bin/tests/bootstrap-integration.bats:35]: By definition,

```
Unit testing means testing individual modules of an application in isolation (without any interaction with dependencies) to confirm that the code is doing things right. Integration testing means checking if different modules are working fine when combined together as a group.
```

This is testing different modules working as a group. The other file is testing functions in the bootstrap module.

File: tools/Dockerfile (hunk 6 above)
Comment (svatwork) [reply tools/Dockerfile:89]: Ok, will have another look.

File: bin/tests/bootstrap-unit.bats (hunk 3 above)
Comment (svatwork) [reply bin/tests/bootstrap-unit.bats:23]: Ok, will have another look

Comment (j-zimnowoda) [reply bin/tests/bootstrap-unit.bats:23]: This statement could be part of `setup` function.
So before running any test you ensure that environment is clean.

Comment (svatwork) [reply bin/tests/bootstrap-unit.bats:23]: Yes that happens with line 10:

```
test_temp_dir="$(temp_make --prefix 'otomi-values-')"
```

Comment (svatwork) [reply bin/tests/bootstrap-unit.bats:23]: But I realise my mistake, I'm about to push some changes.

**Review by Morriz — APPROVED** [review]
